// @niftyco/config/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _fs = require('fs');

var _path = require('path');

var _lodash = require('lodash');

class Config {
  constructor(dir, env = process.env.NODE_ENV) {
    this.$config = {};
    if (dir) {
      this.$load(dir, env);
    }
  }

  $load(dir, env) {
    return (0, _fs.readdirSync)(dir).filter(file => file.endsWith('.js')).map(file => {
      const name = (0, _path.basename)(file, '.js');
      const envFile = (0, _path.resolve)(dir, env, file);
      const prodObj = require((0, _path.resolve)(dir, file));
      const envObj = (0, _fs.existsSync)(envFile) ? require(envFile) : {};
      this.$config[name] = (0, _lodash.defaultsDeep)(envObj, prodObj);
    });
  }

  all() {
    return this.$config;
  }

  has(path) {
    return (0, _lodash.has)(this.$config, path);
  }

  get(path, value) {
    return (0, _lodash.get)(this.$config, path, value);
  }

  set(path, value) {
    return (0, _lodash.set)(this.$config, path, value);
  }

  unset(path) {
    return (0, _lodash.unset)(this.$config, path);
  }
}
exports.default = Config;
module.exports = exports['default'];
